Java/core/src/main/java/com/amazon/randomcutforest/tree/CompactNodeView.java

```diff
@@ -15,31 +15,48 @@
 
 package com.amazon.randomcutforest.tree;
 
+import static com.amazon.randomcutforest.CommonUtils.checkArgument;
+
 import java.util.Arrays;
 import java.util.Set;
 
-public class CompactNodeView implements INode<Integer> {
-    final AbstractCompactRandomCutTree<?> tree;
+import com.amazon.randomcutforest.store.INodeStore;
+
+public class CompactNodeView<Point> implements INode<Integer> {
+    final AbstractCompactRandomCutTree<Point> tree;
     int currentNodeOffset;
+    IBoxCache<Point> boxCache;
+    INodeStore nodeStore;
 
-    public CompactNodeView(AbstractCompactRandomCutTree<?> tree, int initialNodeIndex) {
+    public CompactNodeView(AbstractCompactRandomCutTree<Point> tree, int initialNodeIndex) {
         this.tree = tree;
         this.currentNodeOffset = initialNodeIndex;
+        boxCache = tree.boxCache;
+        nodeStore = tree.nodeStore;
     }
 
     public int getMass() {
-        return tree.getMass(currentNodeOffset);
+        return nodeStore.getMass(currentNodeOffset);
     }
 
     public IBoundingBoxView getBoundingBox() {
-        return tree.getBoundingBox(currentNodeOffset);
+        return getBox(currentNodeOffset);
+    }
+
+    IBoundingBoxView getBox(int node) {
+        if (!nodeStore.isLeaf(node)) {
+            IBoundingBoxView box = boxCache.getBox(node);
+            if (box != null)
+                return box;
+        }
+        return tree.getBoundingBox(node);
     }
 
     public IBoundingBoxView getSiblingBoundingBox(double[] point) {
```

Comment:
We don't have to change it here, but this name always trips me up because I expect it to return the sibling of the current node. I wonder if there's another name that would be clearer.